design/BufferOperations.md

```diff
@@ -0,0 +1,149 @@
+
+The a mapped buffer can be unmapped with:
+
+```
+partial interface WebGPUBuffer {
+    void unmap();
+};
+```
+
+This operation invalidates all the `WebGPUMappedMemory` created from the buffer and puts the buffer in the unmapped state.
+The buffer must be in the mapped state otherwise a validation error occurs when `unmap` is called.
+
+### WebGPUMappedMemory
+
+`WebGPUMappedMemory` is an object representing a mapped region of a buffer that's both pollable and promise-like.
+
+It can be in one of three states: pending, available and invalidated.
+
+The pollable interface is:
+
+```
+partial interface WebGPUMappedMemory {
+    bool isPending();
+    ArrayBuffer getPointer();
+};
+```
+
+`isPending` return true if the object is in the pending state, false otherwise.
+`getPointer` returns an ArrayBuffer representing the buffer data if the object is in the available state, null otherwise.
+
+`WebGPUMappedMemory` is acts like a promise as in that it is then-able:
+
+```
+partial interface WebGPUMappedMemory {
+    void then(WebGPUMappedMemorySuccessCallback success,
+              optional WebGPUMappedMemoryErrorCallback error);
+};
+```
+
+This acts like a `Promise<ArrayBuffer>.then` that is resolved on the Javascript task boundary in which the implementation detects the GPU is done with the buffer.
+On that boundary:
+
+ - The `WebGPUMappedMemory` goes in the available state.
+ - If the `WebGPUMappedMemory` was created via `WebGPUBuffer.mapWrite`, its content is cleared to 0.
+ - `success` is called with the content of the memory as an argument.
+
+If `success` hasn't been called when the WebGPUMappedMemory gets invalidated (meaning the object is still in the pending state), `error` is called instead.
+
+When `WebGPUMappedMemory` goes from the available state to the invalidated state, the `ArrayBuffer` for its content gets neutered.
+
+## Immediate data upload
+
+Buffer mapping is the path with the least number of copies but it is often useful to upload data to a buffer *right now*, if only for debugging.
+A `WebGPUBuffer` operation is provided that takes an ArrayBuffer and copies its content at an offset in the buffer.
+
+```
+partial interface WebGPUBuffer {
+    void setSubData(ArrayBuffer data, u32 offset);
+}
+```
+
+This operation acts as if it was done after all previous "device-level" commands and before all subsequent "device-level" commands.
+"Device level" commands are all commands not buffered in a `WebGPUCommandBuffer`, and include `WebGPUQueue.submit`.
+The content of `data` is only read during the call and can be modified by the application afterwards.
+The following must be true or a validation error occurs:
+
+ - The buffer must have been created with the `WebGPUBufferUsage.TRANSFER_DST` usage flag.
+ - `offset + data.length` must not overflow and be at most the size of the buffer.
+ - Depending on the design of memory barriers, the buffer must be, or allowed to be in the `WebGPUBufferUsage.TRANSFER_DST` usage.
```

Comment:
similar concern here about the mismatch of usage versus state concepts. Are they merged in NXT?

Reply:
No but they are often confused there's the creation usage and the usage "state" (there's other state like the mapped state). So we never use the word "state" for usage, only "usage".

A resource **is in** a usage means it has that usage "state" value
A resource **has a** usage means it was created with this usage requested by the application.

We obviously we didn't spend enough time figuring out the naming of things so this is a bit confusing. I think the intent is clear here, what do you think of deferring the re-wording to when we decide what our memory barriers look like (if any)?

Reply:
I do see a bit of elegance in merging usages and access flags, but it's also confusing to differentiate is/has...

Reply:
I agree with @kvark.  The distinction confused me as well.  

Can buffers only be in more than one "is in" usage at a time?  If not, then that usage should be a separate type that is not a bitfield.  That makes it clear to the person calling functions that take "is in" usage that only one of them is valid. 

If the set of "has a" usages contains memebers that are not valid for "is in" usage, then that's also a reason to have separate types.